astropy/stats/mypyc_demo.py

```diff
@@ -0,0 +1,63 @@
+from collections.abc import Iterable, Sequence
+
+import numpy as np
+from numpy.typing import NDArray
+
+
+def interval_overlap_length(i1: tuple[float, float], i2: tuple[float, float]) -> float:
+    """Compute the length of overlap of two intervals.
+
+    Parameters
+    ----------
+    i1, i2 : (float, float)
+        The two intervals, (interval 1, interval 2).
+
+    Returns
+    -------
+    l : float
+        The length of the overlap between the two intervals.
+
+    """
+    (a, b) = i1
+    (c, d) = i2
+    if a < c:
+        if b < c:
+            return 0.0
+        return b - c if b < d else d - c
+    if a < d:
+        return b - a if b < d else d - a
+    return 0.0
```

Comment:
tentative branchless version (could be incorrect, I just spent 5min on the back of an envelope):
```suggestion
    return min(b - a, min(c - d, max(b - c, 0.0)))
```

Reply:
That fails the tests.

Reply:
Oh yes, this doesn't respect the symmetry of the problem. I'll try to take another shot later.

Reply:
Ok here's a slightly less ambitious version (still haven't been able to fully test it yet)
```suggestion
    if d - b > 0:
        candidate = b - c
    else:
        candidate = d - a
    return min(b - a, min(c - d, max(candidate, 0.0)))
```

Reply:
I'd appreciate if you would stop making code suggestions that don't work.

Reply:
Sounds good. As a proof of concept there's not so much point micro-optimizing this python. Just thinking it would prevent number of branches to hit to demonstrate speedup in all scenarios. 
I think the general point that mypyc is a major speedup and a good idea in general stands.

Reply:
Sorry, I shouldn't rush to suggestions without testing. Now I did, and I found the bug in my latest suggestion (I inverted `c` and `d` on the last line). This passes a small test suite I wrote:
```suggestion
    if d - b > 0:
        candidate = b - c
    else:
        candidate = d - a
    return min(b - a, min(d - c, max(candidate, 0.0)))
```

Reply:
Also @nstarman : I like ternaries as much as the next guy, but here I'm specifically trying to kill if/else branching, so I don't want to hide what's left of it; it's actually better in the annoyingly verbose, itching form :)

Reply:
> compile to identical bytecode, and I would be very surprised if Mypyc and GCC are worse at optimizing than CPython.

yes, I'm not pretending otherwise. I'm just saying it's good to have branching be obvious so it's clear how much of it is left.

>  It was a factor of two slower than my patch with interpreted Python, so almost a factor of two slower than current main

branchless optimization isn't supposed to accelerate interpeted Python, it's only relevant for compiled code. Unfortunately, as long as I use more than 0 unpredictible branches, it's not very surprising if it doesn't work: pipelines will still get flushed :/

>  What kind of a benchmark did you run?

none yet, I'll get there when/if I can get a 0 branch (correct) version

Reply:
(but I did get massive success with this strategy in previous projects, which is why I'm trying it here)

Reply:
I understand, but I'm also interested to know how mypyc compares to Cython when compiling a branchless function.

and for completion, here's the fully branchless version
```suggestion
    candidate = (d - a) + 0.5 * (1.0 + copysign(1.0, d - b)) * ((b - c) - (d - a))
    return min(b - a, min(d - c, max(candidate, 0.0)))
```